lucet-runtime/lucet-runtime-internals/src/instance/execution.rs

```diff
@@ -217,12 +308,52 @@ impl KillState {
         *self.thread_id.lock().unwrap() = None;
         self.tid_change_notifier.notify_all();
     }
+
+    /// Check if the execution domain is currently locked.
+    ///
+    /// This uses [`Mutex::try_lock`](std/sync/struct.Mutex.html#method.try_lock), to see if
+    /// acquiring a lock of the execution domain's mutex, would block the current thread. If
+    /// so, return `true`. If the mutex is not locked, or is in a poisoned state, return false.
+    ///
+    /// This is used by an [`Instance`](struct.KillState.html)'s signal handler to check that a
+    /// `SIGALRM` received during execution came from a kill switch, rather than some other source.
+    pub(crate) fn domain_is_locked(&self) -> bool {
+        use std::sync::TryLockError::WouldBlock;
+        if let Err(WouldBlock) = self.execution_domain.try_lock() {
+            true
+        } else {
+            false
+        }
+    }
+
+    /// Acquire a lock on the execution domain.
+    ///
+    /// This is used by an [`Instance`](struct.KillState.html)'s signal handler to disable
+    /// termination before switching to the host context, for signals _other than_ a `SIGALRM`
+    /// that came from a kill switch, rather than some other source.
+    pub(crate) fn execution_domain(
+        &self,
+    ) -> std::sync::LockResult<std::sync::MutexGuard<'_, Domain>> {
+        self.execution_domain.lock()
+    }
 }
 
+/// Instance execution domains.
+///
+/// This enum allow us to distinguish what an appropriate mechanism to signal termination is.
+#[derive(Debug, PartialEq)]
 pub enum Domain {
+    /// Represents an instance that is not currently running.
+    /// FIXME KTM 2020-02-20: Maybe we should call this `Ready` or `Paused`.
```

Comment:
Agh, I didn't see this comment here! I responded to this elsewhere, but I think pending is deceptively named because we return to this state after running.